### ISSUE TITLE
Building a self contained, then single file version of app will build bad single file version.

### ISSUE THREAD

**Issue by samsp-msft** [dotnet/sdk#12871]
In working on a blog post for single file, I have been building a set of different publish versions of the app. I had been deleting the bin and obj folders between each trim mode, but not between self contained and single file. The latter failed if the bin & obj had not been cleaned first.
It seems like dotnet publish should include a clean as part of its build steps - if it can't handle what is already there, then it needs to do a clean. It should not fail if the existing state is not what it expected.

Failure occurs on windows and Linux (in WSL 2).

```
rm -f -r -d bin
rm -f -r -d obj
dotnet publish -r linux-x64 -p:PublishTrimmed=True -c Release -p:TrimMode=CopyUsed   -o ../AssemblyTrimmed
# rm -f -r -d bin
# rm -f -r -d obj
dotnet publish -r linux-x64 -p:PublishTrimmed=True -c Release -p:TrimMode=CopyUsed   -p:PublishSingleFile=true  -o ../AssemblyTrimmedSF  
```
The latter will succeed in publishing but fail at runtime with 
```
A fatal error was encountered. The library 'libhostpolicy.so' required to execute the application was not found in '/usr/share/dotnet'.
Failed to run as a self-contained app.
  - The application was run as a self-contained app because '/home/sam/FullTrimSF/hello_test.runtimeconfig.json' was not found.
  - If this should be a framework-dependent app, add the '/home/sam/FullTrimSF/hello_test.runtimeconfig.json' file and specify the appropriate framework.
```
if the second set of rm are commented out.

**Comment by agocke** [comment]
I dug into this, and the problem seems to be around the `_CreateAppHost` and `_ChooseAppHost` targets. In short, the `_ChooseAppHost` takes in a bunch of properties and then decides which apphost to use, and sets the path to the correct one. Then, `_CreateAppHost` deploys the chosen apphost to the appropriate path.

The problem is that the dependencies for the targets end up just being the apphost file in the NuGet package, and the outputs are the apphost on disk. In an incremental scenario, if you change the properties on the command line to pick a different app host, the file which will get picked changes, but the up-to-date check only relies on timestamp, not file content, so regardless of which one gets chosen the file is considered up-to-date (because whichever one was written is certainly older than the file in the NuGet directory).

@dsplaisted @rainersigwald I know this is kind of a long-standing limitation in MSBuild, but I don't know what the best way to resolve this is. Effectively, we need the output of `_ChooseAppHost` to play a role in whether or not the targets are up-to-date. Maybe we could write this path to a file, but that seems like a lot of infrastructure to add to the targets.

**Comment by samsp-msft** [comment]
Could the apphost on disk be named differently based on the arguments to the publish (form a hash or something) so that the name of the output is different based on the arguments. It will then always be able to pick the right apphost across builds (assuming it doesn't just pack everything in the output folder).

**Comment by wli3** [comment]
We could use hash property then write down to a file for timestamp compare to fix this incremental build issue.

Publish historical has bad incremental build support. I think this issue existed for a long time. I think we should fix it in 6.0 time instead of 5.0 @marcpopMSFT @dsplaisted @sfoslund

**Comment by samsp-msft** [comment]
I disagree on postponing this to .NET 6 - it produces broken output, with no clear indication that is likely to have happened. From what I can tell, there is not an option as part of publish as to whether to do an incremental build or not, it just happens. 
In its current state, if you do any experimentation, then you are likely to end up with broken output and no indication as to why. Publish should include a clean by default, or incremental publishing needs to be fixed.

@richlander @KathleenDollard

**Comment by wli3** [comment]
I see. Unfortunately we focus on resolving shipping blocker for these months. I'll put this higher in the list of bugs we want to resolve. However we have this issue since 1.0. It would be harder to resolve. And it will carry more risk. I cannot promise it could get in net5.0 considering we are in QB mode.

**Comment by agocke** [comment]
I'm going to try the suggestion of outputting the single-file host with a different name. That should mitigate the biggest item.

**Comment by marcpopMSFT** [comment]
Echoing wli, we are raising the bar for what fixes we are taking and focusing on 5.0 scenarios so this is almost certainly going to slip to 6.0.

**Comment by samsp-msft** [comment]
Thanks Andy.
